File: src/bot/handlers.py
from telegram import Update
from telegram.ext import ContextTypes
from telegram.error import TelegramError

from src.utils.logger import setup_logger

logger = setup_logger()


async def post_message_to_channel(
    context: ContextTypes.DEFAULT_TYPE, channel_id: str, message: str
) -> None:
    try:
        await context.bot.send_message(chat_id=channel_id, text=message)
    except TelegramError as exc:
        logger.error("Telegram API error while posting to channel: %s", exc)
        raise


async def global_error_handler(
    update: object, context: ContextTypes.DEFAULT_TYPE
) -> None:
    logger.error("Unhandled exception: %s", context.error)
    if isinstance(update, Update) and update.effective_chat:
        try:
            await context.bot.send_message(
                chat_id=update.effective_chat.id,
                text="Something went wrong. Please try again later.",
            )
        except TelegramError as exc:
            logger.error("Failed to send error message to user: %s", exc)

    # Future integration point:
    # - OpenAI request/response orchestration hooks.
    # - Supabase persistence/retrieval hooks.

File: src/bot/handlers.ts
import type { Context, Telegraf } from 'telegraf';
import { TelegramError } from 'telegraf';

import type { AppConfig } from '../config.js';
import { logger } from '../utils/logger.js';

export async function postMessageToChannel(
  bot: Telegraf<Context>,
  channelId: string,
  message: string,
): Promise<void> {
  try {
    await bot.telegram.sendMessage(channelId, message);
  } catch (error: unknown) {
    if (error instanceof TelegramError) {
      logger.error(`Telegram API error while posting to channel: ${error.description}`);
    } else {
      logger.error('Unknown error while posting to channel', error);
    }
    throw error;
  }
}

export function registerEventHandlers(bot: Telegraf<Context>, config: AppConfig): void {
  bot.catch((error: unknown, ctx) => {
    const updateType = ctx.updateType;

    if (error instanceof TelegramError) {
      logger.error(`Telegram API error on update [${updateType}]: ${error.description}`);
      return;
    }

    logger.error(`Unhandled bot error on update [${updateType}]`, error);
  });

  // Future place for non-command handlers (text, callback queries, etc.)
  // - OpenAI response generation can be plugged in here.
  // - Supabase read/write logic can be added in repository modules.
  bot.on('message', async (ctx, next) => {
    logger.info(`Incoming message from chat=${ctx.chat.id}`);
    await next();
  });

  logger.info(`Handlers registered (target channel: ${config.telegramChannelId})`);
}
